kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt

```diff
@@ -0,0 +1,318 @@
+      request = requestType.toSchemaType().dataRef,
+      response = responseType.toSchemaType().dataRef,
+      metadata = metadata,
+      comments = verb.comments(),
+    )
+
+    decls.addAll(
+      listOf(
+        Decl(verb = verb),
+        Decl(data_ = requestType.toSchemaData()),
+        Decl(data_ = responseType.toSchemaData())
+      )
+    )
+
+    return ModuleData(decls = decls, comments = module.comments())
+  }
+
+  private fun extractIngress(): MetadataIngress? {
+    return verb.annotationEntries.firstOrNull {
+      bindingContext.get(BindingContext.ANNOTATION, it)?.fqName?.asString() == Ingress::class.qualifiedName
+    }?.let {
+      val argumentLists = it.valueArguments.partition { arg ->
+        // Method arg is named "method" or is of type xyz.block.ftl.Method (in the case where args are
+        // positional rather than named).
+        arg.getArgumentName()?.asName?.asString() == "method"
+          || arg.getArgumentExpression()?.getType(bindingContext)?.fqNameOrNull()
+          ?.asString() == Method::class.qualifiedName
+      }
+      val methodArg = requireNotNull(argumentLists.first.single().getArgumentExpression()?.text) {
+        "Could not extract method from ${verb.name} @Ingress annotation"
+      }
+      // NB: trim leading/trailing double quotes because KtStringTemplateExpression.text includes them
+      val pathArg = requireNotNull(argumentLists.second.single().getArgumentExpression()?.text?.trim { it == '\"' }) {
+        "Could not extract path from ${verb.name} @Ingress annotation"
+      }
+      MetadataIngress(
+        path = pathArg,
+        method = methodArg,
+      )
+    }
+  }
+
+  private fun extractCalls(): MetadataCalls? {
+    val verbs = mutableListOf<VerbRef>()
+    extractCalls(verb, verbs)
+    return verbs.ifNotEmpty { MetadataCalls(calls = verbs) }
+  }
+
+  private fun extractCalls(func: KtNamedFunction, calls: MutableList<VerbRef>) {
+    val body = requireNotNull(func.bodyExpression) { "Verbs must have a body" }
+    val imports = func.containingKtFile.importList?.imports?.mapNotNull { it.importedFqName } ?: emptyList()
+
+    val refs = callMatcher.findAll(body.text).map {
+      val req = requireNotNull(it.groups["req"]?.value?.trim()) {
+        "Could not extract request type for outgoing verb call from ${verb.name}"
+      }
+      val verbCall = requireNotNull(it.groups["fn"]?.value?.trim()) {
+        "Could not extract module name for outgoing verb call from ${verb.name}"
+      }
+      // TODO(worstell): Figure out how to get module name when not imported from another Kt file
+      val moduleRefName = imports.filter { it.toString().contains(req) }.firstOrNull()?.moduleName()
+
+      VerbRef(
+        name = verbCall.split("::")[1].trim(),
+        module = moduleRefName ?: "",
+      )
+    }
+    calls.addAll(refs)
+
+    // Step into function calls inside this expression body to look for transitive calls.
+    body.children.mapNotNull {
```

Comment:
Is this recursive, or just top-level calls? If the latter, fine for now but maybe add a note to remind us.

Reply:
it's recursive!